File: tmp_audit/audit_test08.js
const fs = require('fs');
const path = require('path');
const zlib = require('zlib');

const root = path.join(process.cwd(), 'tmp_audit/test08/levels/test08/art');
const daePath = path.join(root, 'road/road_surface.dae');
const terPath = path.join(root, 'terrains/terrain.ter');
const daeText = fs.readFileSync(daePath, 'utf8');
const terBytes = fs.readFileSync(terPath);
const size = terBytes.readUInt32LE(1);
const sampleCount = size * size;
const heightScale = 500.0 / 65535.0;
const heightU16 = [];
for (let i = 0; i < sampleCount; i++) {
  heightU16.push(terBytes.readUInt16LE(5 + i * 2));
}
function sampleTerrain(x, y) {
  const column = x / 1.0;
  const row = (size - 1) - y / 1.0;
  const c0 = Math.min(size - 2, Math.floor(column));
  const r0 = Math.min(size - 2, Math.floor(row));
  const c1 = c0 + 1;
  const r1 = r0 + 1;
  const tx = column - c0;
  const ty = row - r0;
  const z00 = heightU16[r0 * size + c0] * heightScale;
  const z10 = heightU16[r0 * size + c1] * heightScale;
  const z01 = heightU16[r1 * size + c0] * heightScale;
  const z11 = heightU16[r1 * size + c1] * heightScale;
  const z0 = z00 + (z10 - z00) * tx;
  const z1 = z01 + (z11 - z01) * tx;
  return z0 + (z1 - z0) * ty;
}
const posMatch = daeText.match(/<float_array id="RoadMesh-positions-array" count="(\d+)">([^<]+)<\/float_array>/);
const posVals = posMatch[2].trim().split(/\s+/).map(Number);
const positions = [];
for (let i = 0; i < posVals.length; i += 3) {
  positions.push([posVals[i], posVals[i + 1], posVals[i + 2]]);
}
const pMatch = daeText.match(/<p>([^<]+)<\/p>/s);
const pVals = pMatch[1].trim().split(/\s+/).map(Number);
const triangles = [];
for (let i = 0; i < pVals.length; i += 9) {
  triangles.push([pVals[i], pVals[i + 3], pVals[i + 6]]);
}
const stationCount = positions.length / 7;
const crownIndex = 3;
const crownPositions = [];
for (let s = 0; s < stationCount; s++) {
  crownPositions.push(positions[s * 7 + crownIndex]);
}
let maxCandXY = 0;
let maxCandZ = 0;
for (let i = 1; i < crownPositions.length; i++) {
  const a = crownPositions[i - 1];
  const b = crownPositions[i];
  const dxy = Math.hypot(b[0] - a[0], b[1] - a[1]);
  const dz = Math.abs(b[2] - a[2]);
  if (dxy > maxCandXY) maxCandXY = dxy;
  if (dz > maxCandZ) maxCandZ = dz;
}
let maxEdge = 0;
let maxZExtent = 0;
let edgeGt3 = 0;
let zExtentGt1 = 0;
let firstSep = null;
for (let triIndex = 0; triIndex < triangles.length; triIndex++) {
  const tri = triangles[triIndex];
  const pts = tri.map((idx) => positions[idx]);
  for (let i = 0; i < 3; i++) {
    const a = pts[i];
    const b = pts[(i + 1) % 3];
    const edge = Math.hypot(b[0] - a[0], b[1] - a[1]);
    if (edge > maxEdge) maxEdge = edge;
    if (edge > 3.0) {
      edgeGt3 += 1;
      if (!firstSep) firstSep = { type: 'edge', triIndex, edge };
    }
  }
  const zExtent = Math.max(...pts.map((p) => p[2])) - Math.min(...pts.map((p) => p[2]));
  if (zExtent > maxZExtent) maxZExtent = zExtent;
  if (zExtent > 1.0) {
    zExtentGt1 += 1;
    if (!firstSep) firstSep = { type: 'zextent', triIndex, zExtent };
  }
}
let firstMismatch = null;
for (let s = 0; s < stationCount; s++) {
  for (let v = 0; v < 7; v++) {
    const p = positions[s * 7 + v];
    const terrainZ = sampleTerrain(p[0], p[1]);
    const delta = p[2] - terrainZ;
    if (Math.abs(delta) > 0.29) {
      firstMismatch = { station: s, vertex: v, dae: p, terrain: terrainZ, delta };
      break;
    }
  }
  if (firstMismatch) break;
}

const candidates = [
  ['no_flip', (x, y) => [x, y]],
  ['x_flip', (x, y) => [1023 - x, y]],
  ['y_flip', (x, y) => [x, 1023 - y]],
  ['xy_flip', (x, y) => [1023 - x, 1023 - y]],
];
for (const [label, fn] of candidates) {
  let worst = null;
  let maxAbsDelta = 0;
  for (let s = 0; s < stationCount; s++) {
    for (let v = 0; v < 7; v++) {
      const p = positions[s * 7 + v];
      const [qx, qy] = fn(p[0], p[1]);
      const terrainZ = sampleTerrain(qx, qy);
      const delta = p[2] - terrainZ;
      if (Math.abs(delta) > maxAbsDelta) {
        maxAbsDelta = Math.abs(delta);
        worst = { station: s, vertex: v, emitted: p, sampled: [qx, qy], terrain: terrainZ, delta };
      }
    }
  }
  console.log(label, 'max_abs_delta_to_candidate_terrain', maxAbsDelta.toFixed(6), 'worst', JSON.stringify(worst));
}
console.log('max_crown_xy_distance', maxCandXY.toFixed(6));
console.log('max_crown_z_difference', maxCandZ.toFixed(6));
console.log('max_triangle_edge_length', maxEdge.toFixed(6));
console.log('max_triangle_vertical_extent', maxZExtent.toFixed(6));
console.log('triangles_edge_gt_3m', edgeGt3);
console.log('triangles_zextent_gt_1m', zExtentGt1);
console.log('first_sep', JSON.stringify(firstSep));
console.log('first_mismatch', JSON.stringify(firstMismatch));
console.log('dae_minmax', { minX: Math.min(...positions.map((p) => p[0])), maxX: Math.max(...positions.map((p) => p[0])), minY: Math.min(...positions.map((p) => p[1])), maxY: Math.max(...positions.map((p) => p[1])), minZ: Math.min(...positions.map((p) => p[2])), maxZ: Math.max(...positions.map((p) => p[2])) });
console.log('terrain_decoded_minmax_z', { min: Math.min(...heightU16) * heightScale, max: Math.max(...heightU16) * heightScale });

File: tmp_audit/audit_test08.py
import re, math, struct
from pathlib import Path

root = Path('tmp_audit/test08/levels/test08/art')
text = (root / 'road/road_surface.dae').read_text(encoding='utf-8')
arr = re.search(r'<float_array id="RoadMesh-positions-array" count="(\d+)">([^<]+)</float_array>', text)
vals = [float(x) for x in arr.group(2).split()]
positions = [vals[i:i+3] for i in range(0, len(vals), 3)]
tri_match = re.search(r'<triangles[^>]*count="(\d+)".*?<p>([^<]+)</p>', text, re.S)
tri_vals = [int(x) for x in tri_match.group(2).split()]
triangles = [tri_vals[i:i+3] for i in range(0, len(tri_vals), 3)]

ter_path = root / 'terrains/terrain.ter'
data = ter_path.read_bytes()
version = data[0]
size = int.from_bytes(data[1:5], 'little')
sample_count = size * size
offset = 5
height_u16 = list(struct.unpack('<%dH' % sample_count, data[offset:offset + 2 * sample_count]))
height_scale = 500.0 / 65535.0


def sample_terrain(x, y):
    column = x / 1.0
    row = (size - 1) - y / 1.0
    if column < 0 or column > size - 1 or row < 0 or row > size - 1:
        raise ValueError((x, y, column, row))
    c0 = min(size - 2, math.floor(column))
    r0 = min(size - 2, math.floor(row))
    c1 = c0 + 1
    r1 = r0 + 1
    tx = column - c0
    ty = row - r0
    z00 = height_u16[r0 * size + c0] * height_scale
    z10 = height_u16[r0 * size + c1] * height_scale
    z01 = height_u16[r1 * size + c0] * height_scale
    z11 = height_u16[r1 * size + c1] * height_scale
    z0 = z00 + (z10 - z00) * tx
    z1 = z01 + (z11 - z01) * tx
    return z0 + (z1 - z0) * ty

station_count = len(positions) // 7
crown_idx = 3
crown_positions = [positions[s * 7 + crown_idx] for s in range(station_count)]

print('size', size)
print('station_count', station_count)
print('positions', len(positions))
print('triangles', len(triangles))
print('terrain_minmax', min(height_u16) * height_scale, max(height_u16) * height_scale)

max_xy = 0.0
max_zdiff = 0.0
for i in range(1, station_count):
    prev = crown_positions[i - 1]
    curr = crown_positions[i]
    d = math.hypot(curr[0] - prev[0], curr[1] - prev[1])
    dz = abs(curr[2] - prev[2])
    max_xy = max(max_xy, d)
    max_zdiff = max(max_zdiff, dz)
print('max_crown_xy_distance', max_xy)
print('max_crown_z_diff', max_zdiff)

max_edge = 0.0
max_zextent = 0.0
edge_bad = 0
z_bad = 0
first_sep = None
for tri in triangles:
    pts = [positions[i] for i in tri]
    for i in range(3):
        a = pts[i]
        b = pts[(i + 1) % 3]
        d = math.dist((a[0], a[1]), (b[0], b[1]))
        max_edge = max(max_edge, d)
        if d > 3.0:
            edge_bad += 1
            if first_sep is None:
                first_sep = ('edge', tri, d)
    zext = max(p[2] for p in pts) - min(p[2] for p in pts)
    max_zextent = max(max_zextent, zext)
    if zext > 1.0:
        z_bad += 1
        if first_sep is None:
            first_sep = ('zext', tri, zext)
print('max_triangle_edge_length', max_edge)
print('max_triangle_z_extent', max_zextent)
print('triangles_edge_gt_3m', edge_bad)
print('triangles_zextent_gt_1m', z_bad)
print('first_sep', first_sep)

# DAE min/max
xs = [p[0] for p in positions]
ys = [p[1] for p in positions]
zs = [p[2] for p in positions]
print('dae_minmax_xy_z', min(xs), max(xs), min(ys), max(ys), min(zs), max(zs))
print('terrain_decoded_minmax_z', min(height_u16) * height_scale, max(height_u16) * height_scale)

# first station/vertex that differs from terrain by >0.5m and first station with big separation
for s in range(station_count):
    for v in range(7):
        p = positions[s * 7 + v]
        try:
            tz = sample_terrain(p[0], p[1])
        except Exception:
            continue
        delta = p[2] - tz
        if abs(delta) > 0.5 and 'first_large' not in locals():
            print('first_large_delta', s, v, p, tz, delta)
            first_large = True
        if abs(delta) > 1.0 and 'first_big' not in locals():
            print('first_big_delta', s, v, p, tz, delta)
            first_big = True
    if 'first_big' in locals() and 'first_large' in locals():
        break

# compare against candidate flips
span = (size - 1) * 1.0
candidates = [
    ('no_flip', lambda x, y: (x, y)),
    ('x_flip', lambda x, y: (span - x, y)),
    ('y_flip', lambda x, y: (x, span - y)),
    ('xy_flip', lambda x, y: (span - x, span - y)),
]
for label, fn in candidates:
    max_delta = 0.0
    worst = None
    for s in range(station_count):
        for v in range(7):
            p = positions[s * 7 + v]
            qx, qy = fn(p[0], p[1])
            try:
                tz = sample_terrain(qx, qy)
            except Exception:
                continue
            delta = p[2] - tz
            if abs(delta) > max_delta:
                max_delta = abs(delta)
                worst = (s, v, p[0], p[1], qx, qy, p[2], tz, delta)
    print(label, 'max_abs_delta_to_candidate_terrain', max_delta, 'worst', worst)

# Compare emitted DAE coordinate to terrain at same coordinates and under flipped coords.
for label, fn in candidates:
    max_xy_dist = 0.0
    worst_xy = None
    for s in range(station_count):
        for v in range(7):
            p = positions[s * 7 + v]
            qx, qy = fn(p[0], p[1])
            try:
                tz = sample_terrain(qx, qy)
            except Exception:
                continue
            dist = math.hypot(p[0] - qx, p[1] - qy)
            if dist > max_xy_dist:
                max_xy_dist = dist
                worst_xy = (s, v, p[0], p[1], qx, qy, dist)
    print(label, 'max_xy_distance_to_candidate_point', max_xy_dist, 'worst', worst_xy)
PY